### ISSUE TITLE
Running a recipe results in error

### ISSUE THREAD

**Issue by psijkof** [OrchardCMS/OrchardCore#2329]
SqliteException: SQLite Error 6: 'database table is locked'.

When running a recipe in the admin

**Comment by psijkof** [comment]
`System.AggregateException: One or more errors occurred. (SQLite Error 6: 'database table is locked'.) ---> Microsoft.Data.Sqlite.SqliteException: SQLite Error 6: 'database table is locked'.

   at Microsoft.Data.Sqlite.SqliteException.ThrowExceptionForRC(Int32 rc, sqlite3 db)

   at Microsoft.Data.Sqlite.SqliteCommand.ExecuteReader(CommandBehavior behavior)

   at Microsoft.Data.Sqlite.SqliteCommand.ExecuteScalar()

   at System.Data.Common.DbCommand.ExecuteScalarAsync(CancellationToken cancellationToken)

--- End of stack trace from previous location where exception was thrown ---

   at Dapper.SqlMapper.ExecuteScalarImplAsync[T](IDbConnection cnn, CommandDefinition command) in C:\projects\dapper\Dapper\SqlMapper.Async.cs:line 1217

   at YesSql.Session.UpdateEntityAsync(Object entity)

   at YesSql.Session.CommitAsync()

   --- End of inner exception stack trace ---

   at System.Threading.Tasks.Task.Wait(Int32 millisecondsTimeout, CancellationToken cancellationToken)

   at YesSql.Session.Dispose()

   at Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngineScope.Dispose()

   at OrchardCore.Hosting.ShellBuilders.ShellContext.ServiceScopeWrapper.Dispose() in C:\projects\orchard2\src\OrchardCore\OrchardCore.Environment.Shell.Abstractions\Builders\ShellContext.cs:line 219

   at OrchardCore.Recipes.Services.RecipeExecutor.ExecuteStepAsync(RecipeExecutionContext recipeStep) in C:\projects\orchard2\src\OrchardCore\OrchardCore.Recipes.Implementations\Services\RecipeExecutor.cs:line 190

   at OrchardCore.Recipes.Services.RecipeExecutor.ExecuteAsync(String executionId, RecipeDescriptor recipeDescriptor, Object environment) in C:\projects\orchard2\src\OrchardCore\OrchardCore.Recipes.Implementations\Services\RecipeExecutor.cs:line 108

   at OrchardCore.Recipes.Services.RecipeExecutor.ExecuteAsync(String executionId, RecipeDescriptor recipeDescriptor, Object environment) in C:\projects\orchard2\src\OrchardCore\OrchardCore.Recipes.Implementations\Services\RecipeExecutor.cs:line 126

   at OrchardCore.Recipes.Services.RecipeExecutor.ExecuteAsync(String executionId, RecipeDescriptor recipeDescriptor, Object environment)

   at OrchardCore.Recipes.Controllers.AdminController.Execute(String basePath, String fileName) in C:\projects\orchard2\src\OrchardCore.Modules\OrchardCore.Recipes\Controllers\AdminController.cs:line 100

   at lambda_method(Closure , Object )

   at Microsoft.Extensions.Internal.ObjectMethodExecutorAwaitable.Awaiter.GetResult()

   at Microsoft.AspNetCore.Mvc.Internal.ActionMethodExecutor.TaskOfActionResultExecutor.Execute(IActionResultTypeMapper mapper, ObjectMethodExecutor executor, Object controller, Object[] arguments)

   at System.Threading.Tasks.ValueTask`1.get_Result()

   at Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker.InvokeActionMethodAsync()

   at Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker.InvokeNextActionFilterAsync()

   at Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker.Rethrow(ActionExecutedContext context)

   at Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)

   at Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker.InvokeInnerFilterAsync()

   at Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.InvokeNextResourceFilter()

   at Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.Rethrow(ResourceExecutedContext context)

   at Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)

   at Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.InvokeFilterPipelineAsync()

   at Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.InvokeAsync()

   at Microsoft.AspNetCore.Builder.RouterMiddleware.Invoke(HttpContext httpContext)

   at Microsoft.AspNetCore.StaticFiles.StaticFileMiddleware.Invoke(HttpContext context)

   at SixLabors.ImageSharp.Web.Middleware.ImageSharpMiddleware.Invoke(HttpContext context)

   at Microsoft.AspNetCore.Diagnostics.StatusCodePagesMiddleware.Invoke(HttpContext context)

   at Microsoft.AspNetCore.Authentication.AuthenticationMiddleware.Invoke(HttpContext context)

   at Microsoft.AspNetCore.StaticFiles.StaticFileMiddleware.Invoke(HttpContext context)

   at OrchardCore.Modules.ModularTenantRouterMiddleware.Invoke(HttpContext httpContext) in C:\projects\orchard2\src\OrchardCore\OrchardCore.Modules\ModularTenantRouterMiddleware.cs:line 82

   at OrchardCore.Modules.ModularTenantContainerMiddleware.Invoke(HttpContext httpContext) in C:\projects\orchard2\src\OrchardCore\OrchardCore.Modules\ModularTenantContainerMiddleware.cs:line 99

   at Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddleware.Invoke(HttpContext context)

---> (Inner Exception #0) Microsoft.Data.Sqlite.SqliteException (0x80004005): SQLite Error 6: 'database table is locked'.

   at Microsoft.Data.Sqlite.SqliteException.ThrowExceptionForRC(Int32 rc, sqlite3 db)

   at Microsoft.Data.Sqlite.SqliteCommand.ExecuteReader(CommandBehavior behavior)

   at Microsoft.Data.Sqlite.SqliteCommand.ExecuteScalar()

   at System.Data.Common.DbCommand.ExecuteScalarAsync(CancellationToken cancellationToken)

--- End of stack trace from previous location where exception was thrown ---

   at Dapper.SqlMapper.ExecuteScalarImplAsync[T](IDbConnection cnn, CommandDefinition command) in C:\projects\dapper\Dapper\SqlMapper.Async.cs:line 1217

   at YesSql.Session.UpdateEntityAsync(Object entity)

   at YesSql.Session.CommitAsync()<---
`

**Comment by Skrypt** [comment]
We need a recipe file that we can use to repro this if you can provide one.

**Comment by psijkof** [comment]
Cannot reproduce. Might have been caused by db browser for sql lite locking the file. Closing this.

**Comment by psijkof** [comment]
It is actually quite easy to reproduce and there seems to be an issue running a recipe from the admin.

When I create a vanilla web project in VS, reference cms.targets, update startup to AddOrchardCms() and UseOrchardCore, create a default site with SqlLite (or SQL Server), I create a tenant with SQL Lite running the setup SaaS recipe. 

Then once the tenant is initialized, from the admin, I chose to run Blank Site recipe, and the behavior I see (tried in multiple tenants), is:

With Sql Lite, the exception above gets thrown,

With SQL Server (localdb in my case), the recipe finally finished, but the server logged several timeout exceptions, like:

```
Exception has been thrown by the target of an invocation. ---> System.Data.SqlClient.SqlException: Timeout expired.  The timeout period elapsed prior to completion of the operation or the server is not responding.
WebApplication6> The statement has been terminated. ---> System.ComponentModel.Win32Exception: The wait operation timed out
WebApplication6>    --- End of inner exception stack trace ---
WebApplication6>    at System.Data.SqlClient.SqlConnection.OnError(SqlException exception, Boolean breakConnection, Action`1 wrapCloseInAction)
WebApplication6>    at System.Data.SqlClient.SqlInternalConnection.OnError(SqlException exception, Boolean breakConnection, Action`1 wrapCloseInAction)
WebApplication6>    at System.Data.SqlClient.TdsParser.ThrowExceptionAndWarning(TdsParserStateObject stateObj, Boolean callerHasConnectionLock, Boolean asyncClose)
WebApplication6>    at System.Data.SqlClient.TdsParser.TryRun(RunBehavior runBehavior, SqlCommand cmdHandler, SqlDataReader dataStream, BulkCopySimpleResultSet bulkCopyHandler, TdsParserStateObject stateObj, Boolean& dataReady)
WebApplication6>    at System.Data.SqlClient.SqlCommand.FinishExecuteReader(SqlDataReader ds, RunBehavior runBehavior, String resetOptionsString)
WebApplication6>    at System.Data.SqlClient.SqlCommand.RunExecuteReaderTds(CommandBehavior cmdBehavior, RunBehavior runBehavior, Boolean returnStream, Boolean async, Int32 timeout, Task& task, Boolean asyncWrite, SqlDataReader ds)
WebApplication6>    at System.Data.SqlClient.SqlCommand.RunExecuteReader(CommandBehavior cmdBehavior, RunBehavior runBehavior, Boolean returnStream, TaskCompletionSource`1 completion, Int32 timeout, Task& task, Boolean asyncWrite, String method)
WebApplication6>    at System.Data.SqlClient.SqlCommand.InternalExecuteNonQuery(TaskCompletionSource`1 completion, Boolean sendToPipe, Int32 timeout, Boolean asyncWrite, String methodName)
WebApplication6>    at System.Data.SqlClient.SqlCommand.ExecuteNonQuery()
WebApplication6>    at YesSql.Services.LinearBlockIdGenerator.LeaseRange(ISession session, String dimension)
WebApplication6>    at YesSql.Services.LinearBlockIdGenerator.GetNextId(ISession session, String dimension)
WebApplication6>    at YesSql.Session.Save(Object entity)
WebApplication6>    at OrchardCore.ContentManagement.ContentDefinitionManager.UpdateContentDefinitionRecord() in C:\projects\orchard2\src\OrchardCore\OrchardCore.ContentManagement\ContentDefinitionManager.cs:line 291
WebApplication6>    at OrchardCore.Contents.Migrations.Create() in C:\projects\orchard2\src\OrchardCore.Modules\OrchardCore.Contents\Migrations.cs:line 22
WebApplication6>    --- End of inner exception stack trace ---
WebApplication6>    at System.RuntimeMethodHandle.InvokeMethod(Object target, Object[] arguments, Signature sig, Boolean constructor, Boolean wrapExceptions)
WebApplication6>    at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
WebApplication6>    at OrchardCore.Data.Migration.DataMigrationManager.UpdateAsync(String featureId) in C:\projects\orchard2\src\OrchardCore\OrchardCore.Data\Migration\DataMigrationManager.cs:line 187
WebApplication6> fail: OrchardCore.Data.Migration.DataMigrationManager[0]
```

I haven't tried this on the oc repo, just used nuget packages (myget beta3 latest) so far.

**Comment by psijkof** [comment]
Also note that using SQL server, the admin reports the recipe was run successful, but opening Content Items or Queries throws another exception

```
ObjectDisposedException: Cannot access a disposed object.
Object name: 'IServiceProvider'.
```

**Comment by jtkech** [comment]
Fixed ;)

- The first issue was in `RecipeResult` which defines an `Id` so updates don't create new documents and we were not able to create a new one with `_recipeStore.CreateAsync(result)` in `RecipeExecutor`. Just removing the `Id` in `RecipeResult` fixed the sqlite locking issue.

- But then, as described above with sql server, i got an `ObjectDisposedException` when opening `Content Items `. I fixed this 2nd issue in the controller action by reloading the shell after the recipe.

      await _shellHost.ReloadShellContextAsync(_shellSettings);

  Note: i want to better check this 2nd issue.

Maybe there are other issues but at least it now works with the scenario described by @psijkof when using sqlite, i will check it with sql server.
